File: testers/sqlite/java/src/java/testdb/TestInsertTest.java
package testdb;

import static org.junit.Assert.*;

import java.util.Random;
import org.junit.Test;

/**
 * TestInsert generates random data for required columns; this verifies that the SqliteAdapter arm
 * in ComputedTestInserts produces values that actually round-trip through every supported table.
 */
public class TestInsertTest {
  private final TestInsert testInsert = new TestInsert(new Random(1172222373));

  @Test
  public void testCustomersInsert() {
    SqliteTestHelper.run(
        c -> {
          var row = testInsert.Customers().insert(c);
          assertNotNull(row);
          assertNotNull(row.customerId());
          assertNotNull(row.name());
        });
  }

  @Test
  public void testCustomersWithCustomization() {
    SqliteTestHelper.run(
        c -> {
          var row = testInsert.Customers().with(r -> r.withName("Custom Name")).insert(c);
          assertNotNull(row);
          assertEquals("Custom Name", row.name());
        });
  }

  @Test
  public void testDepartmentsInsert() {
    SqliteTestHelper.run(
        c -> {
          var row = testInsert.Departments().insert(c);
          assertNotNull(row);
          assertNotNull(row.deptCode());
          assertNotNull(row.deptRegion());
          assertNotNull(row.deptName());
        });
  }

  @Test
  public void testProductsInsert() {
    SqliteTestHelper.run(
        c -> {
          var row = testInsert.Products().insert(c);
          assertNotNull(row);
          assertNotNull(row.productId());
          assertNotNull(row.sku());
          assertNotNull(row.name());
          assertNotNull(row.price());
        });
  }

  @Test
  public void testAllScalarTypesInsert() {
    SqliteTestHelper.run(
        c -> {
          var row = testInsert.AllScalarTypes().insert(c);
          assertNotNull(row);
          assertNotNull(row.id());
          assertNotNull(row.colNotNull()); // the only required column
        });
  }

  @Test
  public void testEmployeesWithDepartmentFK() {
    SqliteTestHelper.run(
        c -> {
          var dept = testInsert.Departments().insert(c);
          // SQLite's TestInsert API takes the FK id as a required parameter (codegen wires it to
          // the composite-FK constraint on employees → departments).
          var emp = testInsert.Employees(dept.compositeId()).insert(c);

          assertNotNull(emp);
          assertEquals(dept.deptCode(), emp.deptCode());
          assertEquals(dept.deptRegion(), emp.deptRegion());
        });
  }

  @Test
  public void testOrdersWithCustomerFK() {
    SqliteTestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);

          assertNotNull(order);
          assertEquals(customer.customerId(), order.customerId());
        });
  }

  @Test
  public void testMultipleInserts() {
    SqliteTestHelper.run(
        c -> {
          var row1 = testInsert.Customers().insert(c);
          var row2 = testInsert.Customers().insert(c);
          var row3 = testInsert.Customers().insert(c);

          assertNotEquals(row1.customerId(), row2.customerId());
          assertNotEquals(row2.customerId(), row3.customerId());
          assertNotEquals(row1.customerId(), row3.customerId());
        });
  }

  @Test
  public void testInsertWithSeededRandom() {
    // Reproducibility: two TestInsert instances with the same seed produce the same generated name
    // (modulo any sequence-based ID; we compare names, which come from the alphanumeric generator).
    var name1 =
        SqliteTestHelper.apply(c -> new TestInsert(new Random(123)).Customers().insert(c).name());
    var name2 =
        SqliteTestHelper.apply(c -> new TestInsert(new Random(123)).Customers().insert(c).name());
    assertEquals(name1, name2);
  }
}

File: testers/sqlite/scala/src/scala/testdb/TestInsertTest.scala
package testdb

import org.junit.Assert.*
import org.junit.Test

import scala.util.Random

/** Generated TestInsert helpers actually persist random rows via the connection (the Scala variant differs from Java's pure-construction API). Verifies the SqliteAdapter arm in ComputedTestInserts
  * produces values that round-trip.
  */
class TestInsertTest {
  private val testInsert = TestInsert(Random(1169258584))

  @Test def customersInsert(): Unit = withConnection {
    val row = testInsert.Customers()
    assertNotNull(row.customerId)
    assertNotNull(row.name)
  }

  @Test def customersWithCustomization(): Unit = withConnection {
    val row = testInsert.Customers(name = "Custom Name")
    assertEquals("Custom Name", row.name)
  }

  @Test def departmentsInsert(): Unit = withConnection {
    val row = testInsert.Departments()
    assertNotNull(row.deptCode)
    assertNotNull(row.deptRegion)
  }

  @Test def productsInsert(): Unit = withConnection {
    val row = testInsert.Products()
    assertNotNull(row.productId)
    assertNotNull(row.sku)
  }

  @Test def allScalarTypesInsert(): Unit = withConnection {
    val row = testInsert.AllScalarTypes()
    assertNotNull(row.id)
    assertNotNull(row.colNotNull)
  }

  @Test def employeesWithDepartmentFK(): Unit = withConnection {
    val dept = testInsert.Departments()
    val emp = testInsert.Employees(DepartmentsId = dept.compositeId)
    assertEquals(dept.deptCode, emp.deptCode)
    assertEquals(dept.deptRegion, emp.deptRegion)
  }

  @Test def ordersWithCustomerFK(): Unit = withConnection {
    val cust = testInsert.Customers()
    val ord = testInsert.Orders(customerId = cust.customerId)
    assertEquals(cust.customerId, ord.customerId)
  }

  @Test def multipleInserts(): Unit = withConnection {
    val r1 = testInsert.Customers()
    val r2 = testInsert.Customers()
    val r3 = testInsert.Customers()
    assertNotEquals(r1.customerId, r2.customerId)
    assertNotEquals(r2.customerId, r3.customerId)
  }

  @Test def seededRandomReproducible(): Unit = {
    val name1 = withConnection {
      TestInsert(Random(123)).Customers().name
    }
    val name2 = withConnection {
      TestInsert(Random(123)).Customers().name
    }
    assertEquals(name1, name2)
  }
}
